00XX-efficient-msms-for-non-native-pickles-verification.md

```diff
@@ -0,0 +1,483 @@
+1. Implement POC FFA sub-circuit for the modified target proof system. Test and benchmark.
+   - A single wide row implementation according to the (hopefully optimal) algorithm chosen in the previous step.
+1. Implement POC FFEC sub-circuit for the modified target proof system. Test and benchmark.
+   - This will probably be just standard affine addition. Should be less problematic than the previous FFA step.
+1. Implement the MSM algorithm in the circuit suggested above. Test and benchmark.
+   - The MSM algorithm for now can be implemented without folding in mind. That is, the circuit can be wider, and maybe it can pass more data through inputs, and verify only parts of the MSM of smaller MSM sizes. The point is to have the algorithm working.
+1. First milestone: Releasing a reduced many-proofs version of MSM algorithm.
+   - This task suggests implementing our MSM algorithm /without/ folding.
+   - Instead, we try to verify "small MSM" where everything fits within one circuit. Then we can release $l$ (or $3l$, e.g. 32 or more) independent proofs that will verify *parts* of the MSM.
+
+
+Phase 2:
+1. Bring folding with IVC into our variant of Kimchi.
+   - If available, use FF IVC (preferred), otherwise the BN254/Grumpkin cycle (temporary fallback).
+   - Analyze folding and try to use it in a simple circuit with one of the Pasta curve. Must be able to prove and verify a circuit. It is independent of this work.
+1. Implement the MSM algorithm with folding and IVC.
+   - The MSM circuit from the previous step should be now properly split into sections and folded.
+1. Evaluate the system and potentially optimise the algorithms.
+   - It is perhaps more wise to not spend too much time at the earlier stages choosing the most optimal approach. This task is for revisit their performance now when the whole proving system is in place. Swapping an FFA or FFEC algorithm for another (if more optimal one is found) should be much easier at this step.
+
+
+## Drawbacks
+
+1. Cloned Kimchi might diverge with the main variant.
+    - Implementing the whole project in a clone of Kimchi can be problematic to unify with the main version if two diverge significantnly.
+    - Alternatively, one could modify the original Kimchi to support arbitrary number of columns first, support additive lookups, and implement folding. Or, perhaps, at least any of the three. This leads to a more elegant and unified solution and contributes to the main proof system that the company implements.
+    - However this approach becomes a bottleneck for delivering the results, since Kimchi needs to be integrated with Pickles and support backwards compatibility.
+    - The future plan is to generalize existing Kimchi so that we can just *use* it in this project. However at the moment customizability of Kimchi is still WIP, so it is more optimal to start with a clone of Kimchi, and unify them later. Same goes for folding and logups (unless it is easy to reuse them instead of cloning).
```

Comment:
Did you mint a new term in `logups`?

Reply:
Used in https://eprint.iacr.org/2023/1284.pdf to describe additive lookups with `1/(beta+f(X))`, so seems somewhat applicable?